### PR TITLE
perf: reuse sh:in nodeSet

### PR DESCRIPTION
**Pull request by giacomociti**
This change boosts validation of `sh:in` constraints with long lists of allowed values.

The existing implementation recreates the set of allowed nodes again and again, for each value to be validated.
With the proposed change, the node set is created only once.

To appreciate the performance gain, you may try a shapes graph like [this](https://s.zazuko.com/28Jc1EU), having a constraint with a huge list of municipalities. Try to validate a few [observations](https://s.zazuko.com/Rr3eY6)

(remember to add the target declaration in the shapes graph):

```turtle
<https://energy.ld.admin.ch/elcom/electricityprice/shape> <http://www.w3.org/ns/shacl#targetClass> <https://cube.link/Observation>  . ```

### CODE REVIEW

File: src/validators.js [hunk 1]
```diff
@@ -133,8 +134,16 @@ function validateHasValueProperty(context, focusNode, valueNode, constraint) {
...
   const { sh } = context.ns
   const inNode = constraint.getParameterValue(sh.in)
+  if (!constraint.nodeSets) {
+    constraint.nodeSets = new TermMap()
+  }
+  let nodeSet = constraint.nodeSets.get(inNode)
+  if (!nodeSet) {
+    nodeSet = new NodeSet(rdfListToArray(context.$shapes.node(inNode)))
+    constraint.nodeSets.set(inNode, nodeSet)
+  }
```
Comment (tpluscode) [inline src/validators.js]: Do we need a list of node sets? I would expect a constraint object to represent a single `sh:in` constraint

```suggestion
  if (!constraint.nodeSet) {
    const inNode = constraint.getParameterValue(sh.in)
    constraint.nodeSet = new NodeSet(rdfListToArray(context.$shapes.node(inNode)))
    constraint.nodeSet = new TermMap()
  }
  const { nodeSet } = constraint
```

**Review by tpluscode — CHANGES_REQUESTED** [review]

File: src/validators.js (hunk 1 above)
Comment (giacomociti) [reply src/validators.js]: my understanding is that all the shapes with a `sh:in` constraint are sharing the same instance

Comment (tpluscode) [reply src/validators.js]: I don't think so. `getParameterValue` gets the `sh:in` from the shape pointer.

Comment (giacomociti) [reply src/validators.js]: correct

**Review by tpluscode — APPROVED** [review]

**Pull request merged**